src/cockpit/channels/filesystem.py

```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
-                suffix = ''.join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789_", k=6))
-                self._temppath = f'{self._path}.cockpit-tmp.{suffix}'
-                try:
-                    fd = os.open(self._temppath, os.O_CREAT | os.O_WRONLY | os.O_EXCL, 0o666)
-                    if stat is not None:
-                        # copy permissions from existing file
-                        os.fchmod(fd, stat.st_mode)
-                        os.fchown(fd, stat.st_uid, stat.st_gid)
-                    break
-                except FileExistsError:
-                    continue
-                except PermissionError as exc:
-                    raise ChannelError('access-denied') from exc
-                except FileNotFoundError as exc:
-                    # directory of path does not exist
-                    raise ChannelError('not-found') from exc
-                except OSError as exc:
-                    raise ChannelError('internal-error', message=str(exc)) from exc
+                with contextlib.suppress(FileNotFoundError):  # delete is idempotent
+                    os.unlink(path)
             else:
-                raise ChannelError('internal-error',
-                                   message=f"Could not find unique file name for replacing {self._path}")
-
-            try:
-                self._tempfile = os.fdopen(fd, 'wb')
-            except OSError:
-                # Should Not Happen™, but let's be safe and avoid fd leak
-                os.close(fd)
-                self.unlink_temppath()
-                raise
-
-        self._tempfile.write(data)
-
-    def do_done(self):
-        if self._tempfile is None:
-            try:
-                os.unlink(self._path)
-            # crash on other errors, as they are unexpected
-            except FileNotFoundError:
-                pass
-        else:
-            self._tempfile.flush()
-
-            if self._tag and self._tag != tag_from_path(self._path):
-                raise ChannelError('change-conflict')
-
-            try:
-                os.rename(self._temppath, self._path)
-            # ensure to not leave the temp file behind
-            except FileNotFoundError as exc:
-                self.unlink_temppath()
-                raise ChannelError('not-found', message=str(exc)) from exc
-            except IsADirectoryError as exc:
-                self.unlink_temppath()
-                # not ideal, but the closest code we have
-                raise ChannelError('access-denied', message=str(exc)) from exc
-            except OSError as exc:
-                self.unlink_temppath()
-                raise ChannelError('internal-error', message=str(exc)) from exc
-
-            self._tempfile.close()
-            self._tempfile = None
+                # otherwise, spool data into a temporary file until EOF then rename into place...
+                with tempfile.NamedTemporaryFile(dir=dirname, prefix=f'.{basename}-', delete=False) as tmp:
```

Comment:
The context `tmp` here looks like it shadows the `tmp` from above which his already in scope. I experimented with this and it seems it actually merely overwrites it, and `tmp` is still available after the `with`. But still, this is hard to follow and unclean.

Reply:
Ya.  I buy that argument.  It was also a bit difficult to convince `mypy` to like this.